pkg/autoscaling/calculation.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoscaling
+
+import (
+	"math"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/v4/server/cluster"
+)
+
+// MetricsTimeDuration is used to get the metrics of a certain time period.
+// TODO: adjust the value or make it configurable.
+var MetricsTimeDuration = 5 * time.Second
+
+func calculate(rc *cluster.RaftCluster, strategy *Strategy) []*Plan {
+	var plans []*Plan
+	var afterQuota uint64
+	if tikvPlans, tikvAfterQuota := getPlans(rc, strategy, TiKV); tikvPlans != nil {
+		plans = append(plans, tikvPlans...)
+		afterQuota += tikvAfterQuota
+	}
+	if tidbPlans, tidbAfterQuota := getPlans(rc, strategy, TiDB); tidbPlans != nil {
+		plans = append(plans, tidbPlans...)
+		afterQuota += tidbAfterQuota
+	}
+	if exceedMaxCPUQuota(strategy, afterQuota) {
+		return nil
+	}
+	return plans
+}
+
+func getPlans(rc *cluster.RaftCluster, strategy *Strategy, component ComponentType) ([]*Plan, uint64) {
```

Comment:
what is `uint64` in return value? the number of scale-out?

Reply:
The total CPU quota of a component if we plan to scale-out.